File: DataStructsAndAlgos/implementations/javascript/src/DataStructures/Graphs/Graph/Graph.js
import { Queue } from "../../Queue/Queue";
import { Stack } from "../../Stack/Stack";
import { LinkedList } from "../../LinkedList/LinkedList";

class Path {
  constructor(node, weight = 0) {
    this.list = new LinkedList();
    this.insertNode(node, Infinity);
    this.weight = weight;
  }
  increment() {
    this.weight += 1;
  }
  incrementBy(val) {
    this.weight += val;
  }
  insertNode(node) {
    this.list.insert(node, Infinity);
  }
  viewCurrent() {
    return this.list.peekRear();
  }
  getWeight() {
    return this.weight;
  }
}

class Edge {
  constructor(node, weight = -1) {
    this.toNode = node;
    this.weight = weight;
  }
}

class GraphNode {
  constructor(elem) {
    this.elem = elem;
    this.marked = false;
    this.children = [];
  }
  mark() {
    this.marked = true;
  }
  unmark() {
    this.marked = false;
  }
}

export class Graph {
  constructor(elem) {
    this.root = null;
    if (!!elem) this.init(new GraphNode(elem));
  }
  init(elem) {
    this.root = new GraphNode(elem);
  }
  build(schema) {
    //append a reference, of the new node, to the schema
    for (let elem in schema) {
      let newNode = new GraphNode(elem);
      if (!!schema[elem].root) this.root = newNode;
      schema[elem]["ref"] = newNode;
    }
    //append the children nodes, of each element in the schema, to the reference node.
    for (let elem in schema) {
      for (
        let childIndex = 0;
        childIndex < schema[elem].connections.length;
        childIndex++
      ) {
        let edgeInformation = schema[elem].connections[childIndex];
        let existingNode = schema[edgeInformation.key].ref;
        schema[elem]["ref"].children.push(
          new Edge(existingNode, edgeInformation.weight)
        );
      }
    }
  }
  updateNode(node, needToMatch) {
    if (needToMatch) return node.mark();
    else return node.unmark();
  }
  traverse(needToMatch, searchFor = -1) {
    let output = [];
    let queue = new Queue();
    queue.push(this.root);
    this.updateNode(this.root, needToMatch);
    let found = false;
    while (queue.length() > 0) {
      let newRoot = queue.pop();
      output.push(parseInt(newRoot.elem, 10));
      for (let i = 0; i < newRoot.children.length; i++) {
        let child = newRoot.children[i].toNode;
        if (parseInt(child.elem, 10) === searchFor) found = true;
        if (child.marked !== needToMatch) {
          this.updateNode(child, needToMatch);
          queue.push(child);
        }
      }
    }
    return { output, found };
  }
  unmarkAll() {
    this.traverse(false);
  }
  DFS(elem) {
    let stack = new Stack();
    stack.push(this.root);
    this.updateNode(this.root, true);
    let found = false;
    while (stack.length() > 0) {
      let newRoot = stack.pop();
      for (let i = 0; i < newRoot.children.length; i++) {
        let child = newRoot.children[i].toNode;
        if (parseInt(child.elem, 10) === elem) found = true;
        if (child.marked === false) {
          this.updateNode(child, true);
          stack.push(child);
        }
      }
    }
    this.unmarkAll();
    return found;
  }
  BFS(elem) {
    let found = this.traverse(true, elem);
    this.unmarkAll();
    return found.found;
  }
  print() {
    let graphArr = this.traverse(true);
    this.unmarkAll();
    return graphArr.output;
  }
  shortestPathTo(elem) {
    let queue = new Queue();
    queue.push(new Path(this.root, 0));
    this.updateNode(this.root, true);
    while (queue.length() > 0) {
      let currentPath = queue.pop();
      let lastNodeInCurrentPath = currentPath.viewCurrent();
      if (parseInt(lastNodeInCurrentPath.elem, 10) === parseInt(elem, 10))
        return currentPath.getWeight() + 1;
      for (let i = 0; i < lastNodeInCurrentPath.children.length; i++) {
        let child = lastNodeInCurrentPath.children[i].toNode;
        if (child.marked === false) {
          this.updateNode(child, true);
          let updatedPath = Object.create(currentPath);
          updatedPath.insertNode(child);
          updatedPath.increment();
          queue.push(updatedPath);
        }
      }
    }
    return -1;
  }
}

File: DataStructsAndAlgos/implementations/python/src/ds/Graphs/graph.py

from .. LinkedList.linkedlist import LinkedList
from ..Queue.queue import Queue
from ..Stack.stack import Stack


class DirectedEdge:
    def __init__(self, toNode, weight=1):
        self.toNode = toNode
        self.weight = weight


class UndirectedEdge:
    def __init__(self, fromNode, toNode, weight=1):
        self.fromNode = fromNode
        self.toNode = toNode
        self.weight = weight


class Path:  # Represents a path, represented as a linked list, that has a weight to get from the beginning node to the end node
    def __init__(self, startingNode, weight=0):
        startingNode.mark()
        self.list = LinkedList(startingNode)
        self.weight = weight

    def updateWeight(self, newWeight):
        self.weight += newWeight

    def insertNode(self, newNode, weight=1):
        self.list.insertAtRear(newNode)
        self.updateWeight(weight)

    def currentLocOnPath(self):
        return self.list.peek().elem

    def getWeight(self):
        return self.weight

    def printPath(self):
        graphList = self.list.printList()
        elementList = []
        # TODO: Refactor this so that it is independent of ints
        for currElem in graphList:
            elementList.append(int(currElem.elem))
        return elementList


class GraphNode:
    def __init__(self, startingElem):
        self.elem = startingElem  # Element
        self.children = []  # An array of EDGES
        self.marked = False

    def push(self, newNode):
        self.children.append(newNode)

    def mark(self):
        self.marked = True

    def unmark(self):
        self.marked = False


class Graph:
    def __init__(self, schema=None):
        if schema == None:
            self.root = None
        else:
            self.build(schema)

    def build(self, schema):
        for vertex in schema:
            newNode = GraphNode(vertex)
            schema[vertex]['obj'] = newNode
            if 'root' in schema[vertex]:
                self.root = newNode
        for vertex in schema:
            props = schema[vertex]
            connections = props['connections']
            for edge in connections:
                props['obj'].push(DirectedEdge(
                    schema[edge['key']]['obj'], edge['weight']))

    def search(self, searchType, elemToFind):
        traverseStruct = False
        if searchType == 'BFS':
            traverseStruct = Queue(self.root)
        elif searchType == 'DFS':
            traverseStruct = Stack(self.root)
        else:
            raise Exception('Invalid Search Method Provided!')
        while traverseStruct.getLength() > 0:
            newRoot = traverseStruct.pop()
            if newRoot.elem == elemToFind:
                return True
            for child in newRoot.children:
                traverseStruct.push(child)
        return False

    def updateNode(self, node, needToMatch):
        if node.marked != needToMatch and needToMatch == True:
            node.mark()
        else:
            node.unmark()

    def traverse(self, needToMatch):
        queue = Queue(self.root)
        self.updateNode(self.root, needToMatch)
        while queue.getLength() > 0:
            newRoot = queue.pop()
            for edge in newRoot.children:
                child = edge.toNode
                if child.marked != needToMatch:
                    self.updateNode(child, needToMatch)
                    queue.push(child)

    def cleanUp(self):
        self.traverse(False)

    def displayGraphInBFSOrder(self):
        queue = Queue(self.root)
        self.root.mark()
        finalStruct = []
        while queue.getLength() > 0:
            newRoot = queue.pop()
            finalStruct.append(int(newRoot.elem))
            for edge in newRoot.children:
                child = edge.toNode
                if child.marked == False:
                    child.mark()
                    queue.push(child)
        self.cleanUp()
        return finalStruct
